pytorch_lightning/callbacks/batch_size_finder.py

```diff
@@ -0,0 +1,423 @@
+from pytorch_lightning.utilities.distributed import rank_zero_info
+from pytorch_lightning.utilities.exceptions import _TunerExitException, MisconfigurationException
+from pytorch_lightning.utilities.memory import garbage_collection_cuda, is_oom_error
+from pytorch_lightning.utilities.parsing import lightning_getattr, lightning_hasattr, lightning_setattr
+from pytorch_lightning.utilities.warnings import rank_zero_warn
+
+
+class BatchSizeFinder(Callback):
+    def __init__(
+        self,
+        mode: str = "power",
+        steps_per_trial: int = 3,
+        init_val: int = 2,
+        max_trials: int = 25,
+        batch_arg_name: str = "batch_size",
+    ) -> None:
+        """Callback try to find the largest batch size for a given model that does not give an out of memory (OOM)
+        error. It works with both training and evalation. All you need to do is add it as a callback inside Trainer
+        and call ``trainer.fit/validate/test/predict()``. Internally it calls the respective step function
+        ``steps_per_trial`` times for each batch size until one of the batch size generates and OOM error.
+
+        Args:
+            mode: search strategy to update the batch size:
+
+                - ``'power'``: Keep multiplying the batch size by 2, until we get an OOM error.
+                - ``'binsearch'``: Initially keep multiplying by 2 and after encountering an OOM error
+                    do a binary search between the last successful batch size and the batch size that failed.
+
+            steps_per_trial: number of steps to run with a given batch size.
+                Ideally 1 should be enough to test if a OOM error occurs,
+                however in practice a few are needed.
+
+            init_val: initial batch size to start the search with.
+
+            max_trials: max number of increase in batch size done before
+               algorithm is terminated
+
+            batch_arg_name: name of the attribute that stores the batch size.
+                It is expected that the user has provided a model or datamodule that has a hyperparameter
+                with that name. We will look for this attribute name in the following places
+
+                - ``model``
+                - ``model.hparams``
+                - ``trainer.datamodule`` (the datamodule passed to the tune method)
+        """
+        supported_modes = ("power", "binsearch")
+        mode = mode.lower()
+        if mode not in supported_modes:
+            raise MisconfigurationException(f"`mode` should be one of {supported_modes}")
+
+        self.mode = mode
+        self.steps_per_trial = steps_per_trial
+        self.init_val = init_val
+        self.max_trials = max_trials
+        self.batch_arg_name = batch_arg_name
+        self.optimal_batch_size = init_val
+
+        self._early_exit = False
+
+        from pytorch_lightning.loggers.base import LightningLoggerBase
+
+        class _BatchSizeFinderDumpedParams(TypedDict):
```

Comment:
Mind taking this out of the init method and use a dataclass. "LightningLoggerBase" can be imported under if TYPE_CHECKING: